distributions/aws/infra_configs/scripts/aws/cognito.py

```diff
@@ -0,0 +1,148 @@
+        try:
+            response = self.cognito_client.create_user_pool(PoolName=self.userpool_name)
+        except ClientError:
+            logger.exception(
+                f"Failed to create userpool {self.userpool_name} in {self.region}"
+            )
+            raise
+        else:
+            self.userpool_id = response["UserPool"]["Id"]
+            self.arn = response["UserPool"]["Arn"]
+            logger.info(
+                f"created userpool: {self.userpool_name} with id: {self.userpool_id}, arn: {self.arn}"
+            )
+            return self.arn
+
+    def create_userpool_domain(self) -> str:
+        try:
+            response = self.cognito_client.create_user_pool_domain(
+                Domain=self.userpool_domain,
+                UserPoolId=self.userpool_id,
+                CustomDomainConfig={"CertificateArn": self.domain_cert_arn},
+            )
+        except ClientError:
+            logger.exception(f"failed to create custom domain {self.userpool_domain}")
+            raise
+        else:
+            self.cloudfront_domain = response["CloudFrontDomain"]
+            logger.info(
+                f"created custom domain {self.userpool_domain} with cloudfront url: {self.cloudfront_domain}"
+            )
+            return self.cloudfront_domain
+
+    def describe_userpool_domain(self) -> dict:
+        try:
+            response = self.cognito_client.describe_user_pool_domain(
+                Domain=self.userpool_domain
+            )
+        except ClientError:
+            logger.exception(
+                f"failed to describe userpool domain {self.userpool_domain}"
+            )
+            raise
+        else:
+            return response
+
+    def create_userpool_client(self, client_name: str, callback_urls: list) -> str:
+        try:
+            response = self.cognito_client.create_user_pool_client(
+                UserPoolId=self.userpool_id,
+                ClientName=client_name,
+                CallbackURLs=callback_urls,
+                AllowedOAuthFlows=["code"],
+                AllowedOAuthScopes=[
+                    "email",
+                    "openid",
+                    "aws.cognito.signin.user.admin",
+                    "profile",
+                ],
+            )
+        except ClientError:
+            logger.exception(
+                f"failed to create user pool client {client_name} in {self.arn}"
+            )
+            raise
+        else:
+            self.client_id = response["UserPoolClient"]["ClientId"]
+            logger.info(f"created app client: {client_name} with id: {self.client_id}")
+            return self.client_id
+
+    def get_domain_status(self) -> str:
+        return self.describe_userpool_domain()["DomainDescription"]["Status"]
+
+    def wait_for_domain_status(
+        self,
+        expected_status: str = "ACTIVE",
+        wait_periods: int = 30,
+        period_length: int = 30,
+    ):
```

Comment:
How are we coming up with this number ?

Reply:
these defaults are worst-case based on my test or documentation. for e.g. in this case the Cognito UI prints a note saying this domain might take up to 15 mins to be active. I never saw it taking that long but this could be increased